### ISSUE TITLE
Epic: exp init

### ISSUE THREAD

**Issue by dberenbaum** [iterative/dvc#6446]
Parent issue for lightweight command for initializing experiments.

Linked issues:

- [ ] #6441 
- [ ] #6442 
- [ ] #6443 
- [ ] #6444 
- [ ] #6445

**Comment by karajan1001** [comment]
Maybe we can have a label for these summary?

**Comment by skshetry** [comment]
From the @dberenbaum's comment on https://github.com/iterative/dvc/pull/6681#issuecomment-928180859.

>How do you anticipate this (dvclive scenario) working for non-interactive mode?
>
> I'm still wondering whether the template option is needed (solely as a cli option, not as an actual template; it could be renamed) to separate the dvclive scenario and provide flexibility to add other scenarios in the future. What do you think?


One way may be to make `--live` mandatory for dvclive scenario unless `live` is set on the config (though there's a question about what to do if `metrics` and `plots` are also specified in the config).

Another would be to create a `--template` or something similar flag (`--with={live, etc.}`)?

**Comment by dberenbaum** [comment]
> Another would be to create a `--template` or something similar flag (`--with={live, etc.}`)?

Although it adds another option, I like this mainly because it's more explicit. Now that we don't have templates, it's less transparent what the command and various options are doing. Having a separate option for this suggests that it's not just another path to add but actually changes the stage configuration. We can also further explain it in the help text for that option.

On a similar note, at least interactive mode should probably have a confirmation so that users can see what will be added to `dvc.yaml` beforehand. There could be an option to do the same outside of interactive mode (`--dry`?). Not sure if this has been discussed before for `dvc stage add`.

**Comment by skshetry** [comment]
Tasks that remains:
- [ ] Improve UI/CLI, messaging, and prompts
- [ ] Add config support
- [ ] Add tests/QA

**Comment by daavoo** [comment]
Might be worth revisiting https://github.com/iterative/dvcyaml-schema/issues/7 to better support the workflow: 

`dvc exp init` for simple cases / basic bootstrapping -> manually edit `dvc.yaml` for customization?

**Comment by skshetry** [comment]
With dvcyaml-schema, my goal was to have same schema for JSONSchema and validation. We have been using pydantic for JSONSchema generation and voluptuous for validation.

There are few issues with using pydantic:
1. It does not have a good support for validating dictionaries with varying keys. `stages` in dvc.yaml is a dictionary with key being stage name and value being stage information. Pydantic is not very good at representing this. Similarly, `deps`/`outs` are also list of dictionaries.
2. It is twice almost slow as voluptuous, because it uses type-hints.
3. It coerces types (though you can make it stricter). It may be a good thing sometimes, but it does make it quite tricky to make voluptuous-based schema to be compatible with pydantic-based schema.

In short term, I think it's better to move pydantic schema to dvc, and support both of those schema. 

Voluptuous also has negatives, that it's not very actively supported, and is a bit hard to extend to have custom validations
compared to pydantic (whose DX is great). 

Also, looking for `cattrs` is another option. Though migrating to any other libraries may take longer time, as it's quite hard to guarantee compatibilities (we rarely test unhappy paths for dvc.yaml) and need for QA'ing.

**Comment by dberenbaum** [comment]
We should probably include docs work also.

**Comment by dberenbaum** [comment]
I noticed a couple of workflow issues when doing some QA of `dvc exp init`:

1. It's unclear when or what to `git add`. If I don't `git add` the `dvc.yaml` or the dependencies, `dvc exp init` works in the workspace but not in temp.
2. If I am adding a data dependency, I most likely want to `dvc add` it rather than `git add`.

**Comment by dberenbaum** [comment]
>     * deps from `exp init` should get `.dvc` files (`dvc add`ed)

We might just start with a hint here. Doing a `dvc add` automatically might be unexpected and make `dvc exp init` much slower. I don't love the hint idea, but it's an easy starting point. 🤔 



> `exp run --temp` should just automatically stage any untracked `.dvc` or `dvc.yaml` files (it seems like this is probably the desired default behavior in all cases, not just `exp init` specific ones?)
> 
>     * this doesn't address everything and there may be other untracked files that users actually need to manually stage for their experiment, but we can't really automate behavior for non-dvc-related files

👍 I also think we could add `dvc.yaml`?

**Comment by skshetry** [comment]
Till now, we have considered dependency to be a git-tracked or dvc-tracked dependency, so we do not track them by default. But makes sense to do for `data` in `exp init`.

Would it be confusing to create `data.dvc` and `dvc.yaml` files?

**Comment by dberenbaum** [comment]
> Would it be confusing to create data.dvc and dvc.yaml files?

Ideally, we could specify in `dvc.yaml` that `data` is a DVC-tracked dependency, but I think it's probably easier to add `data.dvc` than to figure out new `dvc.yaml` syntax for now, unless you have an idea to do that easily. Adding `data.dvc` might be awkward, but it also introduces a basic concept to users. What do you think?

> I don't see any issues with doing this by default (it's only dvc.yaml/.gitignore right?).

👍 

> @dberenbaum, should we also `dvc init` by default on `dvc exp init`?

👍

**Comment by daavoo** [comment]
Just wanted to upvote:

> Use train as a default name? (exp init: iterate on ui, tree and show to-be-generated yaml content #6750 (comment))

At least for the `--type live` option, having the name of the added stage (`live`) to be the same as one of the output fields feels a little confusing

**Comment by dberenbaum** [comment]
What about naming the types to match the use cases, like classical ml and deep learning? `ml` and `dl`? This might be more user-friendly since it focuses less on the products/features involved, plus `dvclive` might expand to non-deep learning use cases, which makes things even more confusing.

**Comment by karajan1001** [comment]
I'm sorry, I do not quite understand why `live` is a `dl` only thing?

**Comment by dberenbaum** [comment]
Two reasons:

1. The stage includes checkpoints.
2. Dvclive was initially created for "live" monitoring of steps during training.

I would consider these to mostly be useful when doing deep learning, although maybe that's my personal bias.

Also, dvclive may become more of a general-purpose logger, in which case it may be useful in any ml scenario.

**Comment by iesahin** [comment]
My 2c: 

`--type` may be renamed to `--feature` for a more flexible approach. 

```
dvc exp init --feature live --feature some-future-feature
```

may place nicely together then.

**Comment by skshetry** [comment]
@iesahin, `dvc exp init` is targeted more in terms of scenarios rather than features, which we have plenty of in `dvc stage add`.

**Comment by dberenbaum** [comment]
@skshetry Can we go with `dl` or `deep` for now since that's the focus of the scenario? We don't even need to rename the default.

**Comment by dberenbaum** [comment]
> `init` dvc repo if uninitialized
> Add data as a dvc-tracked directory

After discussion, these are being de-prioritized for now since they are not simple changes and are not strictly necessary.

**Comment by skshetry** [comment]
> Can we go with dl or deep for now since that's the focus of the scenario? We don't even need to rename the default.

`deep` does not convey enough information, I think. Is `DL` a common abbreviation?

**Comment by dberenbaum** [comment]
> Is `DL` a common abbreviation?

Yeah, it's not like ML, but it's fairly common. Google deep learning abbreviation.

**Comment by skshetry** [comment]
@dberenbaum, with `--dl`, I don't think `dl` as a default stage name makes sense. Maybe we should go with `train` then?

**Comment by dberenbaum** [comment]
What's the status @skshetry?

**Comment by jorgeorpinel** [comment]
> Raise issue if dvc is initialized with --no-scm or if the project is not git init-ed.
> should we also dvc init by default on dvc exp init?

I think `exp init` could even do both `git/dvc init` if needed (https://github.com/iterative/dvc/issues/7109). But definitely at least the latter 👍

**Comment by jorgeorpinel** [comment]
Also, I noticed `params.yaml` is the only required part of the default project structure. Should `exp init` create an empty one (with some comment on top perhaps) if it's not there? Otherwise it fails.

**Comment by dberenbaum** [comment]
> Also, I noticed `params.yaml` is the only required part of the default project structure (even if it's empty). Should `exp init` create an empty one (with some comment on top perhaps) if it's not there? Otherwise it fails.

@jorgeorpinel See my response in https://github.com/iterative/dvc.org/pull/3071#discussion_r766966189.

**Comment by jorgeorpinel** [comment]
> The params section of dvc.yaml currently expects keys from a file

What about generating a dvc.yaml without `params` if there's no params.yaml file?

**Comment by pmrowla** [comment]
related: https://github.com/iterative/dvc/issues/7130
